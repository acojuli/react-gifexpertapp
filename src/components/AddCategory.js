import React, { useState } from 'react';
import PropTypes from 'prop-types';

// Recibimos la funcion por props
export const AddCategory = ({ setCategories }) => {
    
    const [inputValue, setInputValue] = useState(''); //''

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    } 

    const handleSubmit = (e) => {
        // De esta manera prevenimos el comportamiento del formulario que actualiza todal a pagina
        e.preventDefault();
        // Con trim borramos los espacios que hay antes y desp de un string
        if ( inputValue.trim().length > 2 ) {

            // console.log('Submit hecho')
            setCategories( cats => [ inputValue, ...cats ] );
            setInputValue('');
        }

    }
    
    return (
        <form onSubmit={ handleSubmit }>
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )
}

// Definicion del proptype, si no proveen esta funcion no van a poder usar mi componente
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
