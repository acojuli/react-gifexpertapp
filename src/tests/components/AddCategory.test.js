import { shallow } from 'enzyme';
import { AddCategory } from '../../components/AddCategory';

describe('Test on <AddCategory />', () => {

    const setCategories = jest.fn();
    let wrapper;

    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = shallow( <AddCategory setCategories={ setCategories } /> );


    });


    test('should show correclty', () => {
        expect( wrapper ).toMatchSnapshot();
    })
    
    test('should change the text box', () => {

        const input = wrapper.find('input');
        const value = 'Hola Mundo';

        input.simulate('change', { target: { value } });

        expect( wrapper.find('p').text().trim() ).toBe( value );        
    })
    
    test('should dont post on submit', () => {

        wrapper.find('form').simulate('submit', { preventDefault(){} } );

        expect( setCategories ).not.toHaveBeenCalled();

    })

    test('should call setCategories and clean the text box', () => {

        const value = 'Hola Mundo';

        // 1. simular el inputChange
        wrapper.find('input').simulate('change', { target: { value } });

        // 2. simular el submit 
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        // 3. setCategories se debe haber llamado
        expect( setCategories ).toHaveBeenCalled();
        expect( setCategories ).toHaveBeenCalledTimes(1);
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) );

        // 4. el value del input debe de estar ''
        expect( wrapper.find('input').prop('value') ).toBe('');
        
    })
    
    
    
})
