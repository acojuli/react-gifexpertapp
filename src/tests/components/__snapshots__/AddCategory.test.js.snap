// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <AddCategory /> should show correclty 1`] = `
<form
  onSubmit={[Function]}
>
  <p>
     
     
  </p>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
